File: ruby/23_merge_k_sorted_lists.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

# @param {ListNode[]} lists
# @return {ListNode}


# 合并多个链表，之前一道题是合并2个链表，所以自然想到递归，将多个最终转换为合并两个链表的情况
def merge_k_lists(lists)
	_merge_k_lists_(lists, 0, lists.size-1)
end

private def _merge_k_lists_(lists, left, right)
     return lists[left] unless right - left > 0

     mid = (left + right) / 2
     l = _merge_k_lists_(lists, left, mid)
     r = _merge_k_lists_(lists, mid + 1, right)
     return _merge_two_lists_(l, r)
end

private def _merge_two_lists_(l1, l2)
	return l1 || l2 unless l1 && l2

	l1, l2 = l2, l1 if l1.val > l2.val
	l3, l1 = l1, l1.next

	node = l3
	while l1 && l2
		if l1.val < l2.val
			node.next = l1
			l1 = l1.next
		else 
			node.next = l2
			l2 = l2.next
		end

		node = node.next
	end

	node.next = l1 || l2

	l3
end

File: cpp/23_merge_k_sorted_lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

// 和21题类似，不过本题给的是一个链表数组，让我们将数组中链表进行合并
// 利用最小堆，最小堆的根节点是最小的数，所以每次取堆中最小的节点，插入到新链表，
// 如果该节点有next节点，将next节点放进堆中，直到堆中数据为空。

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {        
        vector<ListNode *> minHeap;
        for (auto list: lists) {
            if (list != nullptr) {
                minHeap.push_back(list);
            }
        }
        make_heap(minHeap.begin(), minHeap.end(), compare);
        
        ListNode *res = new ListNode(-1);
        ListNode *cur = res;
        
        while (!minHeap.empty()) {
            pop_heap(minHeap.begin(), minHeap.end(), compare);
            ListNode *node = minHeap.back();
            minHeap.pop_back();
            
            cur->next = node;
            cur = cur->next;
            
            
            if (node->next) {
                minHeap.push_back(node->next);
                push_heap(minHeap.begin(), minHeap.end(), compare);
            }
        }
        
        return res->next;
    }
private:
    static bool compare(ListNode *p1, ListNode *p2) {
        return p1->val > p2->val;
    }
};